src/fastcdc.rs

```diff
@@ -0,0 +1,281 @@
+use super::Engine;
+use std::default::Default;
+use std::cmp;
+use std::mem;
+use Gear;
+
+pub struct FastCDC {
+    current_chunk_size: u64,
+    gear: Gear,
+}
+
+impl Default for FastCDC {
+    fn default() -> Self {
+        FastCDC {
+            current_chunk_size: 0,
+            gear: Gear::default(),
+        }
+    }
+}
+
+
+impl Engine for FastCDC {
+    type Digest = u64;
+
+    #[inline(always)]
+    fn roll_byte(&mut self, b: u8) {
+        self.gear.roll_byte(b);
+    }
+
+    #[inline(always)]
+    fn digest(&self) -> u64 {
+        self.gear.digest()
+    }
+
+    #[inline]
+    fn reset(&mut self) {
+        self.gear.reset();
+        self.current_chunk_size = 0;
+    }
+}
+
+impl FastCDC {
+    /// Create new FastCDC engine with default chunking settings
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Create new `FastCDC` engine with custom chunking settings
+    ///
+    /// `chunk_bits` is number of bits that need to match in
+    /// the edge condition. `CHUNK_BITS` constant is the default.
+    pub fn new_with_chunk_bits(chunk_bits: u32) -> Self {
+        Self {
+            current_chunk_size: 0,
+            gear: Gear::new_with_chunk_bits(chunk_bits),
+        }
+    }
+
+    /// Find chunk edge using `FastCDC` defaults.
+    ///
+    /// See `Engine::find_chunk_edge_cond`.
+    pub fn find_chunk_edge(&mut self, mut buf: &[u8]) -> Option<(usize, u64)> {
+
+        const DIGEST_SIZE: usize = 64;
+        debug_assert_eq!(
+            mem::size_of::<<Self as Engine>::Digest>() * 8,
+            DIGEST_SIZE
+            );
+
+        const SPREAD_BITS: u32 = 3;
+        const WINDOW_SIZE: usize = 64;
+
+        let min_shift = DIGEST_SIZE as u32 - self.gear.chunk_bits - SPREAD_BITS;
+        let max_shift = DIGEST_SIZE as u32 - self.gear.chunk_bits + SPREAD_BITS;
+        let min_size = (1 << (self.gear.chunk_bits - SPREAD_BITS)) as u64;
+        let ignore_size = min_size - WINDOW_SIZE as u64;
+        let avg_size = (1 << self.gear.chunk_bits) as u64;
+        let max_size = (1 << (self.gear.chunk_bits + SPREAD_BITS)) as u64;
+
+        let mut cur_offset = 0usize;
+
+        loop {
+            debug_assert!(self.current_chunk_size < max_size);
+            debug_assert!(cur_offset < max_size as usize);
+
+            if buf.is_empty() {
+                return None
+            }
+
+            // ignore bytes that are not going to influence the digest
+            if self.current_chunk_size < ignore_size {
+                let skip_bytes = cmp::min(ignore_size  - self.current_chunk_size, buf.len() as u64);
```

Comment:
Doing this `min` outside might perform better. Like here: https://github.com/dswd/zvault/blob/master/chunking/src/fastcdc.rs#L99